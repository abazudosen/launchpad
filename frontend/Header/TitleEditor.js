/* @flow */

import React from 'react';
import Helmet from 'react-helmet';

import './TitleEditor.less';

type TitleEditorProps = {|
  title: ?string,
  canEdit: boolean,
  githubUsername: ?string,
  onSetTitle: (title: string) => any,
|};

export default class TitleEditor extends React.Component {
  props: TitleEditorProps;

  state: {|
    title: string,
  |};

  constructor(props: TitleEditorProps) {
    super();

    this.state = {
      title: props.title || '',
    };
  }

  handleTitleChange = (evt: SyntheticInputEvent) => {
    this.setState({
      title: evt.target.value,
    });
  };

  saveTitle = () => {
    if (this.state.title !== this.props.title) {
      // Only set title if it was actually changed
      this.props.onSetTitle(this.state.title);
    }
  };

  handleKeyPress = (evt: SyntheticInputEvent) => {
    if (evt.key === 'Enter') {
      this.saveTitle();
    }
  };

  renderTitle() {
    if (this.props.title) {
      return this.props.title;
    } else {
      return <span className="Header-Title--untitled">(Untitled)</span>;
    }
  }

  renderUser() {
    if (this.props.githubUsername) {
      return (
        <span>
          by{' '}
          <a
            href={`https://github.com/${this.props.githubUsername}`}
            target="_blank"
          >
            {this.props.githubUsername}
          </a>
        </span>
      );
    } else {
      return null;
    }
  }

  render() {
    if (this.props.canEdit) {
      return (
        <div className="Title-Editor">
          <Helmet>
            <meta
              name="twitter:title"
              content={
                this.state.title === '' ? 'Untitled Pad' : this.state.title
              }
            />
            <meta
              property="og:title"
              content={
                this.state.title === '' ? 'Untitled Pad' : this.state.title
              }
            />
          </Helmet>
          <div
            className="edit-title-button"
            aria-label="Edit Title and Description"
          >
            <span key="icon-edit" className="icon-edit" />
          </div>

          <input
            type="text"
            className="TitleInput"
            placeholder="Untitled"
            value={this.state.title}
            onChange={this.handleTitleChange}
            onBlur={this.saveTitle}
            onKeyPress={this.handleKeyPress}
          />
        </div>
      );
    } else {
      return (
        <div className="Title">
          <div>
            {this.renderTitle()} {this.renderUser()}
          </div>
        </div>
      );
    }
  }
}
