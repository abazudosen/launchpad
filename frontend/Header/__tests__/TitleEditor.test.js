/* @flow */

import React from 'react';
import TitleEditor from '../TitleEditor';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import serializer from 'enzyme-to-json/serializer';

expect.addSnapshotSerializer(serializer);

test('render normal', () => {
  let component = renderer.create(
    <TitleEditor
      title="Test"
      githubUsername="example"
      canEdit={true}
      onSetTitle={() => {}}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();

  component = renderer.create(
    <TitleEditor
      title="Test"
      githubUsername="example"
      canEdit={false}
      onSetTitle={() => {}}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();

  component = renderer.create(
    <TitleEditor
      title={null}
      githubUsername="example"
      canEdit={true}
      onSetTitle={() => {}}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();

  component = renderer.create(
    <TitleEditor
      title={null}
      githubUsername={null}
      canEdit={true}
      onSetTitle={() => {}}
    />,
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test('interaction', () => {
  const onSetTitle = jest.fn();
  const component = mount(
    <TitleEditor
      title="test"
      githubUsername="example"
      canEdit={true}
      onSetTitle={onSetTitle}
    />,
  );
  expect(component).toMatchSnapshot();

  component
    .find('.TitleInput')
    .simulate('change', { target: { value: 'New test' } });
  expect(component).toMatchSnapshot();

  component.find('.TitleInput').simulate('blur');
  expect(onSetTitle.mock.calls.length).toBe(1);

  component
    .find('.TitleInput')
    .simulate('change', { target: { value: 'New test 2' } });
  expect(component).toMatchSnapshot();

  component.find('.TitleInput').simulate('keyPress', { key: 'Enter' });
  expect(onSetTitle.mock.calls.length).toBe(2);
});
