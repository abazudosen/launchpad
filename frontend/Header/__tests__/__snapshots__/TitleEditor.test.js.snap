// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interaction 1`] = `
<TitleEditor
  canEdit={true}
  githubUsername="example"
  onSetTitle={[Function]}
  title="test"
>
  <div
    className="Title-Editor"
  >
    <HelmetWrapper
      encodeSpecialCharacters={true}
    >
      <SideEffect(NullComponent)
        encodeSpecialCharacters={true}
        meta={
          Array [
            Object {
              "content": "test",
              "name": "twitter:title",
            },
            Object {
              "content": "test",
              "property": "og:title",
            },
          ]
        }
      >
        <NullComponent
          encodeSpecialCharacters={true}
          meta={
            Array [
              Object {
                "content": "test",
                "name": "twitter:title",
              },
              Object {
                "content": "test",
                "property": "og:title",
              },
            ]
          }
        />
      </SideEffect(NullComponent)>
    </HelmetWrapper>
    <div
      aria-label="Edit Title and Description"
      className="edit-title-button"
    >
      <span
        className="icon-edit"
      />
    </div>
    <input
      className="TitleInput"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Untitled"
      type="text"
      value="test"
    />
  </div>
</TitleEditor>
`;

exports[`interaction 2`] = `
<TitleEditor
  canEdit={true}
  githubUsername="example"
  onSetTitle={[Function]}
  title="test"
>
  <div
    className="Title-Editor"
  >
    <HelmetWrapper
      encodeSpecialCharacters={true}
    >
      <SideEffect(NullComponent)
        encodeSpecialCharacters={true}
        meta={
          Array [
            Object {
              "content": "New test",
              "name": "twitter:title",
            },
            Object {
              "content": "New test",
              "property": "og:title",
            },
          ]
        }
      >
        <NullComponent
          encodeSpecialCharacters={true}
          meta={
            Array [
              Object {
                "content": "New test",
                "name": "twitter:title",
              },
              Object {
                "content": "New test",
                "property": "og:title",
              },
            ]
          }
        />
      </SideEffect(NullComponent)>
    </HelmetWrapper>
    <div
      aria-label="Edit Title and Description"
      className="edit-title-button"
    >
      <span
        className="icon-edit"
      />
    </div>
    <input
      className="TitleInput"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Untitled"
      type="text"
      value="New test"
    />
  </div>
</TitleEditor>
`;

exports[`interaction 3`] = `
<TitleEditor
  canEdit={true}
  githubUsername="example"
  onSetTitle={[Function]}
  title="test"
>
  <div
    className="Title-Editor"
  >
    <HelmetWrapper
      encodeSpecialCharacters={true}
    >
      <SideEffect(NullComponent)
        encodeSpecialCharacters={true}
        meta={
          Array [
            Object {
              "content": "New test 2",
              "name": "twitter:title",
            },
            Object {
              "content": "New test 2",
              "property": "og:title",
            },
          ]
        }
      >
        <NullComponent
          encodeSpecialCharacters={true}
          meta={
            Array [
              Object {
                "content": "New test 2",
                "name": "twitter:title",
              },
              Object {
                "content": "New test 2",
                "property": "og:title",
              },
            ]
          }
        />
      </SideEffect(NullComponent)>
    </HelmetWrapper>
    <div
      aria-label="Edit Title and Description"
      className="edit-title-button"
    >
      <span
        className="icon-edit"
      />
    </div>
    <input
      className="TitleInput"
      onBlur={[Function]}
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Untitled"
      type="text"
      value="New test 2"
    />
  </div>
</TitleEditor>
`;

exports[`render normal 1`] = `
<div
  className="Title-Editor"
>
  <div
    aria-label="Edit Title and Description"
    className="edit-title-button"
  >
    <span
      className="icon-edit"
    />
  </div>
  <input
    className="TitleInput"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyPress={[Function]}
    placeholder="Untitled"
    type="text"
    value="Test"
  />
</div>
`;

exports[`render normal 2`] = `
<div
  className="Title"
>
  <div>
    Test
     
    <span>
      by
       
      <a
        href="https://github.com/example"
        target="_blank"
      >
        example
      </a>
    </span>
  </div>
</div>
`;

exports[`render normal 3`] = `
<div
  className="Title-Editor"
>
  <div
    aria-label="Edit Title and Description"
    className="edit-title-button"
  >
    <span
      className="icon-edit"
    />
  </div>
  <input
    className="TitleInput"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyPress={[Function]}
    placeholder="Untitled"
    type="text"
    value=""
  />
</div>
`;

exports[`render normal 4`] = `
<div
  className="Title-Editor"
>
  <div
    aria-label="Edit Title and Description"
    className="edit-title-button"
  >
    <span
      className="icon-edit"
    />
  </div>
  <input
    className="TitleInput"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyPress={[Function]}
    placeholder="Untitled"
    type="text"
    value=""
  />
</div>
`;
